python/apps/taiga/src/taiga/users/services/__init__.py

```diff
@@ -234,7 +248,119 @@ async def update_user(user: User, full_name: str, lang: str) -> User:
 #####################################################################
 
 
-# TODO
+async def delete_user(user: User) -> bool:
+    # delete workspaces where the user is the only ws member
+    # (all members, invitations, projects, pj-members, pj-invitations,
+    # pj-roles, stories, comments, etc will be deleted in cascade)
+    workspaces = await workspaces_repositories.list_workspaces(
+        filters={"workspace_member_id": user.id, "num_members": 1}
+    )
+    for ws in workspaces:
+        ws_deleted = await workspaces_repositories.delete_workspaces(filters={"id": ws.id})
+        if ws_deleted > 0:
+            await workspaces_events.emit_event_when_workspace_is_deleted(workspace=ws, deleted_by=user)
+
+    # delete projects where the user is the only pj member
+    # (all members, invitations, pj-roles, stories, comments, etc
+    # will be deleted in cascade)
+    projects = await projects_repositories.list_projects(
+        filters={"project_member_id": user.id, "is_onewoman_project": True},
+    )
+    [await projects_services.delete_project(project=pj, deleted_by=user) for pj in projects]
+
+    # update role of a workspace member as project admin in projects where the user is the only pj admin
+    # better if the workspace member is project member too
+    projects = await projects_repositories.list_projects(
+        filters={"project_member_id": user.id, "is_admin": True, "num_admins": 1, "is_onewoman_project": False},
+        select_related=["workspace"],
+    )
+    for pj in projects:
+        workspace_members = await ws_memberships_repositories.list_workspace_members_excluding_user(
+            workspace=pj.workspace, exclude_user=user
+        )
+        project_members = await pj_memberships_repositories.list_project_members_excluding_user(
+            project=pj, exclude_user=user
+        )
+
+        project_admin_role = await pj_roles_repositories.get_project_role(
+            filters={"project_id": pj.id, "slug": "admin"}
+        )
+        common_members = list(set(workspace_members).intersection(project_members))
+        if len(common_members) > 0:
+            pj_membership = await pj_memberships_repositories.get_project_membership(
+                filters={"project_id": pj.id, "user_id": common_members[0].id},
```

Comment:
I think I've got it. As sets are unordered, getting the first element it's a sort of take any random common member. Anyone will do, is it so?

Reply:
Yes, I think it is a good solution for that :)